### PR TITLE
[KIP-86] Implements rd_kafka_sasl_set_credentials

### PR DESCRIPTION
**Pull request by josvisser66**
This PR adds a new API call rd_kafka_sasl_set_credentials which can be used to set or reset the credentials used for SASL PLAIN or SCRAM in a Kafka client. This is useful for environments that want to change their credentials (e.g. because the old credentials are going to expire) without destroying the Kafka client. The new credentials will be used the next time the Kafka client needs to make a new connection to a broker. Setting these credentials using the new API call will not break any connections that were set up using the old credentials.

Note: Although this is not the same as KIP-86, it allows for the same sort of behavior as KIP-86.

### CODE REVIEW

File: src/rdkafka.c [hunk 1]
```diff
@@ -346,6 +346,29 @@ rd_kafka_resp_err_t rd_kafka_oauthbearer_set_token_failure(rd_kafka_t *rk,
 #endif
 }
 
+rd_kafka_error_t *rd_kafka_set_sasl_creds(rd_kafka_t *rk, const char *username, const char *password) {
+        char errstr[256];
+        rd_kafka_conf_t *conf = &rk->rk_conf; /* Not using rd_kafka_conf because of const qualifier */
+
+        rwlock_wrlock(&conf->sasl.lock);
```
Comment (edenhill) [inline src/rdkafka.c:353]: No need for an rwlock, this is not in the fast-path. Use a regular mtx_t (mtx_plain).

File: .idea/codeStyles/codeStyleConfig.xml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+<component name="ProjectCodeStyleConfiguration">
```
Comment (edenhill) [inline .idea/codeStyles/codeStyleConfig.xml:1]: Please remove this file.

Also make sure to fixup the style by running `make style-fix` in the top-level directory prior to committing.

File: src/rdkafka.c [hunk 3]
```diff
@@ -346,6 +346,29 @@ rd_kafka_resp_err_t rd_kafka_oauthbearer_set_token_failure(rd_kafka_t *rk,
 #endif
 }
 
+rd_kafka_error_t *rd_kafka_set_sasl_creds(rd_kafka_t *rk, const char *username, const char *password) {
+        char errstr[256];
+        rd_kafka_conf_t *conf = &rk->rk_conf; /* Not using rd_kafka_conf because of const qualifier */
```
Comment (edenhill) [inline src/rdkafka.c:351]: I'm not sure I understand this comment, rk_conf is non-const.

File: src/rdkafka.c [hunk 4]
```diff
@@ -346,6 +346,29 @@ rd_kafka_resp_err_t rd_kafka_oauthbearer_set_token_failure(rd_kafka_t *rk,
...
+        if (RD_KAFKA_CONF_OK != rd_kafka_conf_set(conf, "sasl.password", password, errstr, sizeof(errstr))) {
+                goto error;
+        }
+
+        rwlock_wrunlock(&conf->sasl.lock);
+        return NULL;
+
+error:
+        rwlock_wrunlock(&conf->sasl.lock);
+        return rd_kafka_error_new(RD_KAFKA_RESP_ERR__INVALID_ARG, "rd_conf_set failed: %s", errstr);
```
Comment (edenhill) [inline src/rdkafka.c:368]:
```suggestion
        return rd_kafka_error_new(RD_KAFKA_RESP_ERR__INVALID_ARG, "Failed to update SASL credentials: %s", errstr);
```

File: src/rdkafka.c [hunk 5]
```diff
@@ -346,6 +346,29 @@ rd_kafka_resp_err_t rd_kafka_oauthbearer_set_token_failure(rd_kafka_t *rk,
 #endif
 }
 
+rd_kafka_error_t *rd_kafka_set_sasl_creds(rd_kafka_t *rk, const char *username, const char *password) {
```
Comment (edenhill) [inline src/rdkafka.c:349]: I'd move this to rdkafka_sasl.c and rename it to `rd_kafka_sasl_set_credentials()`

File: src/rdkafka.c [hunk 6]
```diff
@@ -2204,6 +2227,7 @@ rd_kafka_t *rd_kafka_new(rd_kafka_type_t type,
         rd_kafka_interceptors_on_new(rk, &rk->rk_conf);
 
         rwlock_init(&rk->rk_lock);
+        rwlock_init(&rk->rk_conf.sasl.lock);
```
Comment (edenhill) [inline src/rdkafka.c:2230]: use mtx, also add mtx_destroy() to destroy_final()

File: src/rdkafka_sasl_plain.c [hunk 7]
```diff
@@ -70,10 +70,14 @@ int rd_kafka_sasl_plain_client_new(rd_kafka_transport_t *rktrans,
                                    size_t errstr_size) {
         rd_kafka_broker_t *rkb = rktrans->rktrans_rkb;
         rd_kafka_t *rk         = rkb->rkb_rk;
+        rd_kafka_conf_t *conf  = &rk->rk_conf; /* Not using rd_kafka_conf because of const modifier */
```
Comment (edenhill) [inline src/rdkafka_sasl_plain.c:73]: Other places in this function already use rk->rk_conf directly, so no need for this alias.

**Review by edenhill — CHANGES_REQUESTED** [review]
Great stuff!

File: src/rdkafka.c (hunk 3 above)
Comment (josvisser66) [reply src/rdkafka.c:351]: const rd_kafka_conf_t *rd_kafka_conf(rd_kafka_t *rk) returns a const* to an rd_kafka_conf_t. The calls to lock/unlock break that const modifier.

File: src/rdkafka.c (hunk 4 above)
Comment (josvisser66) [reply src/rdkafka.c:368]: Done

File: src/rdkafka.c (hunk 5 above)
Comment (josvisser66) [reply src/rdkafka.c:349]: Done

File: src/rdkafka_sasl_plain.c (hunk 7 above)
Comment (josvisser66) [reply src/rdkafka_sasl_plain.c:73]: Done

File: src/rdkafka.c (hunk 3 above)
Comment (edenhill) [reply src/rdkafka.c:351]: No need for rd_kafka_conf(), that's part of the public API. Inside librdkafka you should access rk->rk_conf directly.

File: CHANGELOG.md [hunk 8]
```diff
@@ -1,3 +1,9 @@
+# librdkafka v1.9.4
+
+librdkafka v1.9.4 is a maintenance release:
+
+* Added rd_kafka_sasl_set_credentials API function to set or reset the SASL PLAIN or SCRAM credentials used for new broker connections.
```
Comment (edenhill) [inline CHANGELOG.md:5]: This should go in the v1.9.3 section below - that's the next release.
Add it to the Enhancement chapter
Try to stick below 80 columns

Comment (josvisser66) [reply CHANGELOG.md]: Fixed

**Review by edenhill — APPROVED** [review]

**Comment by edenhill** [comment]
Thanks Jos!

**Pull request merged**